api/v1alpha1/contour_types.go

```diff
@@ -89,6 +89,17 @@ type ContourSpec struct {
 	// +kubebuilder:validation:MaxLength=253
 	// +optional
 	GatewayClassRef *string `json:"gatewayClassRef,omitempty"`
+
+	// IngressClassName is the name of the IngressClass used by Contour. If unset,
```

Comment:
I think there needs to be a bit more subtlety here, sadly. The Contour controller currently defaults to assuming it's the default in the cluster, indicated by an unset or `""` value for the `--ingress-class-name` argument/field.

I think the contract for this field needs to be:
- unset: default for cluster. The operator sets everything up so that Contour is the default (either no IngressClass created or create one and set it as the default, I can't remember which is preferred for IngressClass)
- set: the flag is set to the given value, and an IngressClass is created that fits that contract. It is *not* set as the cluster default.

The value can't default to `contour` because that will imply the second behavior when it's passed to the controller, sadly.

Reply:
> The Contour controller currently defaults to assuming it's the default in the cluster, indicated by an unset or "" value for the --ingress-class-name argument/field.

@youngnick according to the [contour ingressclass docs](https://projectcontour.io/docs/main/config/annotations/#ingress-class), "contour" is the default value of --ingress-class-name. Therefore, I'm having the operator provide the same default behavior, using api defaulting instead of the `ingressClassName` field being a pointer. Are the contour ingressclass docs incorrect regarding "--ingress-class-name=contour" or am I misunderstanding the docs?

Reply:
If unset Contour processes ALL ingress objects without an ingress-class annotation or with an annotation matching `ingress-class=contour`. 

When you set the ingress-class explicitly, the Contour ONLY processes ingress objects that match that class.

Reply:
So the default should be no ingress class, then is someone wants to configure one, they can.

Reply:
So now that this is a point, we have a clear distinction between "unset" and "set", and we can replicate the Contour behavior? If so, looks good to me.

Reply:
@youngnick yes, the latest commit makes `contour.spec.IngressClassName` optional to replicate Contour's default behavior: If unspecified, Contour processes all ingress objects without an ingress-class annotation or with an annotation matching ingress-class=contour. If specified, Contour processes ingress objects with an annotation matching the value of `contour.spec.IngressClassName`. I've updated the description of the PR accordingly.